packages/perseus/src/renderer-util.ts

```diff
@@ -0,0 +1,99 @@
+import Util from "./util";
+import {getWidgetValidator} from "./widgets";
+
+import type {PerseusWidgetsMap} from "./perseus-types";
+import type {PerseusStrings} from "./strings";
+import type {PerseusScore} from "./types";
+import type {UserInput} from "./validation.types";
+
+// It can be nested because of Groups
+// TODO: can it be infinitely nested? How do we type that?
+type UserInputMap = Record<string, UserInput | Record<string, UserInput>>;
+
+export function emptyWidgetsFunctional(
+    widgets: PerseusWidgetsMap,
+    // This is a port of old code, I'm not sure why
+    // we need widgetIds vs the keys of the widgets object
+    widgetIds: Array<string>,
+    userInputMap: UserInputMap,
+    strings: PerseusStrings,
+    locale: string,
+): ReadonlyArray<string> {
+    return widgetIds.filter((id) => {
+        const widget = widgets[id];
+        if (!widget || widget.static) {
+            // Static widgets shouldn't count as empty
+            return false;
+        }
+
+        let score: PerseusScore | null = null;
+        const userInput = userInputMap[id];
+        const validator = getWidgetValidator(widget.type);
+
+        if (widget.type === "group") {
```

Comment:
Another way to approach this would be to have `group` have a validator and have it call back into this function.  That would eliminate the custom code paths for just `group`.  We also have  `graded-group` (and maybe `graded-group-set`) that do similar tricks. Not sure if those need to be addressed here too.